src/test/kotlin/com/atlassian/performance/tools/hardware/HardwareApdexAnalysisIT.kt

```diff
@@ -0,0 +1,269 @@
+            .filter { file ->
+                // ignore any empty run/# folders or ones that only contain a status.txt which indicates a failed run
+                file.listFiles().any{ it.name != "status.txt" }
+            }
+            .map { file ->
+                apdexMetrics(file)
+            }
+            .toList()
+
+        // get the best worst runs as defined by the overall appdex
+        val sortedRuns = runs
+            .sortedByDescending { it.apdex }
+        val best = sortedRuns.first()
+        val worst = sortedRuns.last()
+
+        // comparison
+        compare(best, worst)
+    }
+
+    private fun compare(best: ApdexResults, worst: ApdexResults) {
+        // summary
+        compareRuns(best, worst)
+
+        // detail
+        compareActionEventByLabel(best, worst)
+    }
+
+    private fun compareRuns(best: ApdexResults, worst: ApdexResults) {
+        val table = GenericPlainTextTable()
+
+        // header
+        table.addRow(listOf( "", "Run", "Apdex", "Total Actions", "Error %", "Satisfactory %", "Tolerable %"))
+
+        // data
+        tableAddRunSummaryRow(table, best, "best")
+        tableAddRunSummaryRow(table, worst, "worst")
+
+        println(table.generate())
+    }
+
+    private fun tableAddRunSummaryRow(table: GenericPlainTextTable, results: ApdexResults, name: String) {
+        val eventTot = results.metrics.size
+        val errorPerc = results.metrics.filter { it.actionMetric.result == ActionResult.ERROR }.size * 100.0 / eventTot
+        val satisfactoryPerc = results.metrics.filter { it.weight == 1.0f }.size * 100.0 / eventTot
+        val tolerablePerc = results.metrics.filter { it.weight == 0.5f }.size * 100.0 / eventTot
+        table.addRow(listOf(
+            name,
+            "${results.run}",
+            "${results.apdex}",
+            "$eventTot",
+            String.format("%.3f", errorPerc),
+            String.format("%.3f", satisfactoryPerc),
+            String.format("%.3f", tolerablePerc)
+            ))
+    }
+
+    private fun compareActionEventByLabel(best: ApdexResults, worst: ApdexResults) {
+
+        val table = GenericPlainTextTable()
+
+        // header
+        table.addRow(listOf( "Action", "Run", "Apdex", "Total", "Profile", "Error %", "Satisfactory %", "Tolerable %"))
+
+        // data
+        best.metrics
+            .asSequence()
+            //.map { it.actionMetric }
+            .groupBy { it.actionMetric.label }
+            .asSequence()
+            .sortedBy { it.key }
+            .forEach { entry ->
+                tableAddActionRow(table, best.run, best.metrics.size, entry.key, entry.value)
+
+                //find the equivalent row form thw worst run
+                val worstRow = worst.metrics.asSequence()
+                    .groupBy { it.actionMetric.label }
+                    .asSequence().singleOrNull {
+                        it.key == entry.key
+                    }
+                if(worstRow != null) {
```

Comment:
There are small code-style inconsistencies here and there.
I recommend spamming IntelliJ `Reformat Code` action. I do it all the time.
This way it is not only easier on the eyes, but we're avoiding generating unnecessary diff in future commits.

Reply:
done 👍